exercise/templatetags/exercise.py

```diff
@@ -49,10 +50,25 @@ def _get_toc(context, student=None):
     })
     return context
 
+# Method for setting difficulties for submissions
+# (enabling textual status in points badges).
+def set_difficulties(children):
+    for child in children:
+        if child['submissions']:
+            difficulty = child['difficulty']
+            for submission in child['submissions']:
+                submission['difficulty'] = difficulty
+        elif 'children' in child:
+            set_difficulties(child['children'])
 
 @register.inclusion_tag("exercise/_user_results.html", takes_context=True)
 def user_results(context, student=None):
     values = _get_toc(context, student)
+
+    # Add the difficulty information to all submission to enable
+    # textual status in points badges in submissions list.
+    for module in values['modules']:
+        set_difficulties(module['children'])
```

Comment:
I was reading the code to find why this is actually even needed. What part breaks if this operation is not done?

To me only https://github.com/evnurm/a-plus/blob/ed99c0abe44ec112ca1ac15e51598cc02e827e9e/exercise/templatetags/exercise.py#L256 seems to be possible place where difficulty is now known.

Reply:
![issue](https://user-images.githubusercontent.com/23477304/43244184-79126776-90b2-11e8-80f3-b591ebcab124.png)

I was faced with the issue above: the texts were not showed on the submission list. I assumed that it was because the difficulty information was not configured in the argument given to points_badge. Therefore I decided to add the information this way.

Reply:
Could you test if you can fix it by editing `_points_data`, because I'm pretty sure the problem lies there.

Reply:
It seems like https://github.com/evnurm/a-plus/blob/ed99c0abe44ec112ca1ac15e51598cc02e827e9e/exercise/templatetags/exercise.py#L234 should be executed, for that drop down, thus it should include the difficulty. Thus, I presume the code path goes to else branch due to some reason.

Reply:
I can confirm based on print-testing that the execution always seems to go there - the type of `obj` seems to always be `dict` due to which the `isinstance` conditions evaluate to False.

Reply:
Yeah. Because the data comes from pointscache now-a-days, so the model information is lost. That templatetag should be redesigned to work correctly with pointscache, so it's able to resolve variables correctly. So, `_points_data` should detect the type from dictionary content or we should pass the type as argument from templates.

Reply:
Just realized that those isinstance-blocks do work as intented in the submission_plain and exercise_plain pages (apparently they do not come from cache?). In other words, I think that those should not be changed after all.

I believe all modules, exercises etc. work as intended at the moment in the dict-case so I guess that the easiest fix to this would be to just insert the difficulty information into each submission in the submissions list just the way I had done before. Is there a good reason why it can't be done that way?

Reply:
Unneeded copying of data, that should be resolvable in `_points_data`. Can you check if the dictionary (`obj`) contains reference to the exercise, that way the code could could get it there if the obj it self doesn't contain the info.

Reply:
Does not seem to be the case  - here are the keys: 

['submission_status', 'confirm_the_level', 'graded', 'date', 'passed', 'id', 'url', 'max_points', 'submission_count', 'points_to_pass', 'points', 'unofficial']

Reply:
I could also perform the addition of difficulty in the phase where the cached data is generated so it does not need to be changed later in the template tag logic.  Is that a better idea? I would call that addition instead of copying.

Reply:
I decided to implement my suggestion as I feel that it's better than the previous solution. It's now pushed into this PR.

Reply:
Can you move the change after line `confirm_the_level': entry.get('confirm_the_level', False),`. In addition, it a bit looks like entry is reference to the exercise, so check if that contains `difficulty`.

Reply:
Looks like it does contain `difficulty`. I implemented it that way and moved the change onto that line - should be pushed as well.